src/AutoMapper/ConfigurationStore.cs

```diff
@@ -394,7 +405,8 @@ public TypeMap ResolveTypeMap(TypePair typePair)
                                 FindTypeMapFor(tp) ??
                                 (!CoveredByObjectMap(typePair)
                                     ? FindConventionTypeMapFor(tp) ??
-                                      FindClosedGenericTypeMapFor(tp)
+                                      FindClosedGenericTypeMapFor(tp) ??
+                                      (CreateMissingTypeMaps ? CreateTypeMap(tp) : null)

```

Comment:
And the func passed to GetOrAdd should be thread safe. I don't think it is. CreateMap isn't, see #1018 for example.